### PR TITLE
LibFaadLoader for m4a without recompile 

### PR DESCRIPTION
**Pull request by daschuer**
LibFaadLoader dynamic loads faad2 whenever it is found on the users system. 
This is allows to play m4a files without recompile on Linux. 

It is enabled by default on Linux only, because the other systems have their own solution.

### CODE REVIEW

**Comment by uklotzde** [comment]
Now only dynamic binding is possible. Instead we should introduce an abstract base class and implementations that provide both, i.e. either static or dynamic binding.

Presumably loading of shared libraries at runtime from the host will not be supported by Flatpak or Snap!

**Comment by daschuer** [comment]
The issue here is that we cannot ship a flat pack with libfaad, because it suffers patent issues. 
Only the user can decide if it is possible to use the library or not.

**Comment by daschuer** [comment]
Ok, it is working now. Appveyor is timed out.  

In case we decide to static link libfaad some day, it should be quite easy to add this support. So IMHO there is nothing wrong to merge this. 

The only point we should consider is reusing the FAAD=1 flag. 
The current situation here is now: 
* Users that have build Mixxx with FAAD=1 will not notice a difference. The dynamic load of the soundsource is replaced by dynamic load of the libarary. 
* Users that have build Mixxx with FAAD=0 will also not effected. MP4 disabled as usual.  
* Users on Linux that have not touch the FAAD flag will get the loader. They can decode m4a if the OS provides the libfaad. This is new and gives all freedom to the user without recompile Mixxx. Without this we need to provide two versions of Mixxx depending on the license situation, or exclude some users from the m4a feature even though other Apps on their device are able to do that. 
* Windows and Mac users are not effeted by default because their OS already allows m4a encoding like before. 

Any thoughts?

File: src/sources/libfaadloader.cpp [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+        unsigned long* pSamplerate,
+        unsigned char* pChannels) {
+    if (m_neAACDecInit2) {
+        return m_neAACDecInit2(hDecoder,
+                pBuffer,
+                SizeOfDecoderSpecificInfo,
+                pSamplerate,
+                pChannels);
+    }
+    return 0;
```
Comment (uklotzde) [inline src/sources/libfaadloader.cpp]: -1 would be more appropriate, because 0 is successful.

File: src/sources/libfaadloader.cpp [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    if (m_neAACDecDecode2) {
+        return m_neAACDecDecode2(
+                hDecoder,
+                pInfo,
+                pBuffer,
+                bufferSize,
+                ppSampleBuffer,
+                sampleBufferSize);
+    }
+    return 0;
```
Comment (uklotzde) [inline src/sources/libfaadloader.cpp]: nullptr

File: src/sources/libfaadloader.h [hunk 3]
```diff
@@ -0,0 +1,139 @@
+#pragma once
+
+#include <QLibrary>
+
+#include "util/memory.h"
+#include "util/logger.h"
```
Comment (uklotzde) [inline src/sources/libfaadloader.h]: Move to .cpp

File: src/sources/libfaadloader.h [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+            void** ppSampleBuffer,
+            unsigned long sampleBufferSize);
+
+    char* GetErrorMessage(unsigned char errcode);
+
+
+  private:
+    LibFaadLoader();
+
+    std::unique_ptr<QLibrary> m_pLibrary;
```
Comment (uklotzde) [inline src/sources/libfaadloader.h]: Please add a comment that QLibrary is not copyable. I first thought that the unique_ptr is unecessary here.

**Comment by uklotzde** [comment]
An ugly workaround, but I agree that this is a pragmatic solution after removal of the SoundSource plugins. Since FAAD2 is not actively developed, it should be safe.

File: src/sources/libfaadloader.cpp [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+#elif __WINDOWS__
+    // http://www.rarewares.org/aac-decoders.php
+    libnames << "libfaad2.dll";
+#elif __APPLE__
+    // Using Homebrew ('brew install faad2' command):
+    libnames << "/usr/local/lib/libfaad2.dylib";
+    // Using MacPorts ('sudo port install faad2' command):
+    libnames << "/opt/local/lib/libfaad2.dylib";
+#else
+    DEBUG_ASSERT(!"OS not implementet");
```
Comment (uklotzde) [inline src/sources/libfaadloader.cpp]: typo

File: src/sources/libfaadloader.cpp [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+        m_neAACDecInit2 = nullptr;
+        m_neAACDecClose = nullptr;
+        m_neAACDecPostSeekReset = nullptr;
+        m_neAACDecDecode2 = nullptr;
+        m_neAACDecGetErrorMessage = nullptr;
+        m_pLibrary->unload();
+        m_pLibrary.reset();
+        return;
+    }
+    kLogger.debug() << "Successfully loaded faad2 library " << m_pLibrary->fileName();
```
Comment (uklotzde) [inline src/sources/libfaadloader.cpp]: - Increase log level to info
- Remove trailing space from message string -> "Successfully loaded library" (faad2 will appear as parts of the file name)

File: src/sources/libfaadloader.cpp [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+    // Load shared library
+    QStringList libnames;
+#ifdef __LINUX__
+    libnames << "libfaad.so.2";
+#elif __WINDOWS__
+    // http://www.rarewares.org/aac-decoders.php
+    libnames << "libfaad2.dll";
+#elif __APPLE__
+    // Using Homebrew ('brew install faad2' command):
+    libnames << "/usr/local/lib/libfaad2.dylib";
```
Comment (uklotzde) [inline src/sources/libfaadloader.cpp]: Why not use "libfaad2.dylib" without any path as the first option?

File: src/sources/libfaadloader.cpp (hunk 1 above)
Comment (daschuer) [reply src/sources/libfaadloader.cpp]: No see: 
https://github.com/gypified/libfaad/blob/master/docs/Ahead%20AAC%20Decoder%20library%20documentation.pdf
> Return values:
> 0 – Error, invalid configuration.
> 1 – OK

I will add a comment.

Comment (uklotzde) [reply src/sources/libfaadloader.cpp]: NeAACDecInit2
Return values:
< 0 – Error
0 - OK

**Comment by daschuer** [comment]
Notes addressed, thank you for review.

File: src/sources/libfaadloader.cpp (hunk 1 above)
Comment (daschuer) [reply src/sources/libfaadloader.cpp]: Ups, I have fixed this on the wrong function. Now the new comment is in both places.

**Comment by daschuer** [comment]
Again ready.

**Comment by uklotzde** [comment]
LGTM.

First I was sceptical about this approach, but it is a pragmatic solution that compensates the removal of the SoundSource plugins. Since FAAD2 is EOL and not actively developed I don't expect that we ever need to touch the imported constants and function definitions.

**Pull request merged**